### ISSUE TITLE
Task.onlyIf should provide transitive avoidance

### ISSUE THREAD

**Issue by stephanenicolas** [gradle/gradle#16111]
Gradle <= 6.8.2

Let's say we have tasks with the following dependencies 
```
// the arrows mean A --depends on--> B
A --> B 
       \
         -> E
       /
C --> D
```
If we add an `onlyIf` clause to `B` and this condition is `false`, Gradle should be able to skip both `A` & `B` when invoking task `E`. The current behavior is to only skip `B`, but still execute `A`.
Of course, there are some rare cases where the execution of `A` would influence the `onlyIf` condition on `B`, but they are rare. It would be nice to have an API to allow Gradle to make such a decision. 
e.g.: `B.onlyIfTransitive { <condition> }` <<-- naming is a hard problem :D

**Comment by vlsi** [comment]
@stephanenicolas , would you please use real-looking names instead of A B C D? Names like A B make it really hard to understand the intention

What do the tasks do, what command user issues, and what is the expected result?

**Comment by daymxn** [comment]
> @stephanenicolas , would you please use real-looking names instead of A B C D? Names like A B make it really hard to understand the intention
> 
> What do the tasks do, what command user issues, and what is the expected result?

We're also running into this over at Google.

For example;

We have a `generateReleaseConfig` task which figures out which projects in a multi-module project are opted into a release given some conditions and state.

We additionally have a `publishRelease` task which publishes the release artifacts for a given module. Although, this should only occur if the module is opted into a release (as discovered via `generateReleaseConfig`). Our ideal solution is something like this:
```kotlin
val releasingLibraries = generateReleaseConfig.map { it.releasingLibraries }
val imReleasing = releasingLibraries.map { it.contains(name) }

tasks.register("publishRelease") {
  dependsOn(imReleasing)
  onlyIf { imReleasing.get() }

  finalizedBy("publishReleasePublicationToRepo") // provided by AGP
  finalizedBy(...) // other release tasks
}
```

Although, this doesn't work- as the dependent tasks of `publishRelease` are still ran, even if the `onlyIf` causes the task to skip.

tl;dr -  wrapper tasks that are conditional given some `Provider` are perfect examples of where this becomes an issue.

**Comment by ov7a** [comment]
Sorry for the late reply.

Thank you for your interest in Gradle!

This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

**Comment by lptr** [comment]
The way `Task.onlyIf()` works now is unfortunately broken. As you demonstrated, a downstream task can still consume the excluded task's stale output. The plan is to disallow depending on excluded tasks, though this is not something that we will work on in the near future: https://github.com/gradle/gradle/issues/11100
